PurchasesCoreSwift/Public/PurchaserInfo.swift

```diff
@@ -0,0 +1,326 @@
+
+            // Metadata
+            self.originalApplicationVersion = subscriberData["original_application_version"] as? String ?? nil
+
+            if let originalPurchaseDateString = subscriberData["original_purchase_date"] as? String {
+                self.originalPurchaseDate = PurchaserInfo.parseDate(
+                    dateString: originalPurchaseDateString,
+                    withDateFormatter: PurchaserInfo.dateFormatter)
+            } else {
+                self.originalPurchaseDate = nil
+            }
+
+            guard let firstSeenDateString = subscriberData["first_seen"] as? String,
+                  let firstSeenDate = PurchaserInfo.parseDate(
+                    dateString: firstSeenDateString,
+                    withDateFormatter: dateFormatter) else {
+                return nil
+            }
+            self.firstSeen = firstSeenDate
+
+            guard let originalAppUserIdString = subscriberData["original_app_user_id"] as? String else {
+                return nil
+            }
+            self.originalAppUserId = originalAppUserIdString
+
+            if let managementUrlString = subscriberData["management_url"] as? String {
+                self.managementURL = parseURL(url: managementUrlString)
+            } else {
+                self.managementURL = nil
+            }
+
+            // Purchases and entitlements
+            self.nonSubscriptions =
+                subscriberData["non_subscriptions"] as? [String: [[String: Any]]] ?? [String: [[String: Any]]]()
+            self.entitlements = subscriberData["entitlements"] as? [String: Any] ?? [String: Any]()
+            self.nonSubscriptionTransactions = TransactionsFactory().nonSubscriptionTransactions(
+                withSubscriptionsData: nonSubscriptions,
+                dateFormatter: dateFormatter)
+
+            let latestNonSubscriptionTransactionsByProductId =
+                [String: Any](uniqueKeysWithValues: nonSubscriptions.map { productId, transactionsArray in
+                (productId, transactionsArray.last)
+            })
+
+            self.allPurchases =
+                latestNonSubscriptionTransactionsByProductId.merging(subscriptions) { (current, _) in current }
+
+            self.expirationDatesByProduct = PurchaserInfo.parseExpirationDate(expirationDates: subscriptions)
+            self.purchaseDatesByProduct = PurchaserInfo.parsePurchaseDate(purchaseDates: allPurchases)
+        }
+    }
+
+}
+
+private extension PurchaserInfo {
+
+    func activeKeys(dates: [String: Date?]) -> Set<String> {
+        return Set(dates.keys.filter {
+            guard let nonNullDate = dates[$0] as? Date else { return true }
+            return isAfterReferenceDate(date: nonNullDate)
+        })
+    }
+
+    func isAfterReferenceDate(date: Date) -> Bool { date.timeIntervalSince(self.requestDate) > 0 }
+
+    class func parseURL(url: String) -> URL? { URL(string: url) }
```

Comment:
do these need to be static? 
I usually try to avoid static functions because they're harder to mock. With these being private it shouldn't cause issues though.

Reply:
also, I'm kinda wondering whether having this function makes sense... I'm not sure it adds much clarity: 
```swift
self.managementURL = parseURL(url: managementUrlString)
// vs
self.managementURL = URL(string: url)
```

Reply:
agreed on removing that function... parse date probably isn't much of a help either...

Reply:
as for your first comment..they didn't NEED to be, just thought in general we should make things static if we can?